### DISCUSSION TITLE
memcached-tag

### DISCUSSION THREAD

**Discussion by volodymyr-volynets** [memcached/memcached#1040] [General]
From google code:

We add "Tag Function" for memcached. Propose is remove several keys with the same tag in one operation. This function will help the API programmers (such as php) do the delete operation easily, and reduce the network load. We use hash and splay tree, make the "tag_add" and "tag_delete" commands very quickly, and save memory as much as possible.

**Is there any way we can add tags? Any reply would be appreciated.**

**Comment by dormando** [comment]
Hi.

I can't add tags the way they added them, but I am trying to figure out a way to improve things.

How are you intending to use this feature? How many keys do you need to associate with a particular tag? How often are they deleted?

Most of the time people achieve "tagging" and "Tagged deletion" with namespacing: https://github.com/memcached/memcached/wiki/ProgrammingTricks#namespacing

**Reply by volodymyr-volynets** [reply]
Hi,

Adding tags:

1. Ideally each key can have 0 or up to 50 tags.
2. So we need to have the ability to issue add_tags command with input parameters: key and space separated list of tags 1 or more.
3. Tags cannot contain spaces.

Removing keys with tags:

1. We need to bee able to issue remove_keys_by_tags command with input parameters: space separated list of tags 1 or more.
2. In case of multi-tenancy we can form tag names like tenant_555_table_users

Hope this helps.

**Reply by dormando** [reply]
So you are expecting to be able to delete keys by any number of tags?
IE: if you delete the "user" tag, all user keys are deleted. If you delete the "user555" tag, only keys related to that user are removed?

How are you expecting this to work with a cluster of memcached servers?
IE: there are three servers and your keys are distributed across all of them (see https://memcached.org/tutorial)

**Reply by volodymyr-volynets** [reply]
In terms of multiple servers you execute the same remove by tag command on all servers.

**Reply by volodymyr-volynets** [reply]
And yes you need to be able to delete multiple keys by multiple tags. For example you cache a row from user table with permissions but you update db table of user roles assignment - we need to invalidate those user keys.

**Reply by volodymyr-volynets** [reply]
Hi,

I have DataSources that join 25-50 tables and generate SQL, I generate sha1() from SQL and its my key, tags are tenant number and table names and other objects like functions. When data changes in any of those table I need to invalidate all keys which have random names. Currently I use file based cache and tags are stored in a separate cookie files.

**Reply by dormando** [reply]
None of this works with memcached's design. It probably sounds like it would be okay, but you will immediately run into problems.

- memcached is a _cluster_ (see https://memcached.org/tutorial) - for any key invalidation you would need to issue that command against all servers in the cluster.
- this is a performance problem: memcached's design allows scaling the cache for size or performance reasons by adding more servers. If you add more servers but each server needs to handle all cache invalidation requests, and as you scale you need to _do_ more cache invalidations, you will not be able to scale up.
- this is a management problem: if each invalidation must run successfully on all nodes, how do you guarantee they were all run at once? Do you need to do a two-phase commit to ensure they are all wiped at the same time?

With memcached you will need to adapt your cache pattern to how it works a bit. I'll link this here again and give you some examples: https://github.com/memcached/memcached/wiki/ProgrammingTricks#namespacing

- The simplest starter example is to use a global namespace, and if any table changes, update the namespace and invalidate the cache:
```
get namespace_3_{sha1() hash of whatever} -> hit
update namespace (via incr/set/whatever)
get namespace_4_{hash} -> miss
```
- The next step is to start organizing your cache data. For example, lets namespace against your tenants instead of globally.
```
get tenant_555_3_{sha1()} -> hit
-- any table in the tenant is updated for any reason, update the namespace
get tenant_555_4_{sha1()} -> miss, recache
```
For most people who have patterns like yours, this second example can give a 90%+ hit rate, as most tenants are mostly reads (though I don't know anything about your data). I know it's hard to believe but I have done this exact thing on large social media services.

- From here you can split further, ie: post data vs user data. If post data is modified a lot more often than user data, structure your namespaces so the user data doesn't have to invalidate.

If you cannot easily divide your data down further than that, you need to change your caching model, which is going to be a lot more work and up to you if it makes sense at all. Instead of using SQL to join a bunch of data together, you cache the independent portions of data and join them in the app after fetching from the cache. Then when individual rows of data changes you are specifically invalidating the cache of those objects via standard memcached delete (or using low TTL's where that makes sense instead).

If this doesn't work, memcached isn't a good fit for you. It's a scalable cache, you can't look at its features in terms of a single server.

**Comment by volodymyr-volynets** [comment]
!Bump.
